### PR TITLE
Fix lexical binding set in VM_OC_COPY_TO_GLOBAL

### LINKED ISSUE #3647: Assertion object_p->type_flags_refs >= ECMA_OBJECT_REF_ONE in ecma_deref_object
###### JerryScript revision
655110ad

###### Build platform
Linux-4.15.0-88-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
switch (g) {
    case f():
        let g;
        function f() {
            eval('function g({["y"]: []}) {}; g({xy: {}})')
        }
}

switch (g) {
    case f:
        function f() {}
    case g():
       function g() {}
}
```

###### Output
```text
ICE: Assertion 'object_p->type_flags_refs >= ECMA_OBJECT_REF_ONE' failed at jerryscript/jerry-core/ecma/base/ecma-gc.c(ecma_deref_object):149.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Backtrace
```text
(gdb) bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x5657b570 in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:30
#4  0x56630061 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:63
#5  0x566300a2 in jerry_assert_fail (assertion=0x567091a0 "object_p->type_flags_refs >= ECMA_OBJECT_REF_ONE", file=0x56709140 "jerryscript/jerry-core/ecma/base/ecma-gc.c", 
    function=0x566bfe60 <__func__.6504.lto_priv.910> "ecma_deref_object", line=149) at jerryscript/jerry-core/jrt/jrt-fatals.c:87
#6  0x5669c63c in ecma_deref_object (object_p=0xf5f00490) at jerryscript/jerry-core/ecma/base/ecma-gc.c:149
#7  0x566b45bd in ecma_free_value (value=4126147731) at jerryscript/jerry-core/ecma/base/ecma-helpers-value.c:1047
#8  0x566b4647 in ecma_fast_free_value (value=4126147731) at jerryscript/jerry-core/ecma/base/ecma-helpers-value.c:1076
#9  0x5662397c in vm_loop.lto_priv.447 (frame_ctx_p=0xffffc460) at jerryscript/jerry-core/vm/vm.c:3773
#10 0x565d233f in vm_execute (frame_ctx_p=0xffffc460) at jerryscript/jerry-core/vm/vm.c:4028
#11 0x565d291b in vm_run (bytecode_header_p=0xf5500cd0, this_binding_value=4126149459, lex_env_p=0xf5d003f0, arg_list_p=0x1, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:4136
#12 0x5661113f in vm_run_eval (bytecode_data_p=0xf5500cd0, parse_opts=140) at jerryscript/jerry-core/vm/vm.c:393
#13 0x56648ccf in ecma_op_eval_chars_buffer (code_p=0xf630152c "function g({[\"y\"]: []}) {}; g({xy: {}})", code_buffer_size=39, parse_opts=140)
    at jerryscript/jerry-core/ecma/operations/ecma-eval.c:123
#14 0x56648945 in ecma_op_eval (code_p=0xf6301520, parse_opts=8) at jerryscript/jerry-core/ecma/operations/ecma-eval.c:58
#15 0x565ee3a2 in ecma_builtin_global_object_eval (x=4130346273) at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtin-global.c:111
#16 0x565f1b65 in ecma_builtin_global_dispatch_routine (builtin_routine_id=84, this_arg=72, arguments_list_p=0xffffc8e0, arguments_number=1)
    at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtin-global.c:1162
#17 0x56673125 in ecma_builtin_dispatch_routine (func_obj_p=0xf5f004f0, this_arg_value=72, arguments_list_p=0xffffc8e0, arguments_list_len=1)
    at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtins.c:1056
#18 0x566732fd in ecma_builtin_dispatch_call (obj_p=0xf5f004f0, this_arg_value=72, arguments_list_p=0xffffcb40, arguments_list_len=1)
    at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtins.c:1080
#19 0x5664b306 in ecma_op_function_call_simple (func_obj_p=0xf5f004f0, this_arg_value=72, arguments_list_p=0xffffcb40, arguments_list_len=1)
    at jerryscript/jerry-core/ecma/operations/ecma-function-object.c:751
#20 0x5664c218 in ecma_op_function_call (func_obj_p=0xf5f004f0, this_arg_value=72, arguments_list_p=0xffffcb40, arguments_list_len=1)
    at jerryscript/jerry-core/ecma/operations/ecma-function-object.c:1054
#21 0x56612842 in opfunc_call.lto_priv.446 (frame_ctx_p=0xffffcb10) at jerryscript/jerry-core/vm/vm.c:765
#22 0x565d23a3 in vm_execute (frame_ctx_p=0xffffcb10) at jerryscript/jerry-core/vm/vm.c:4034
#23 0x565d291b in vm_run (bytecode_header_p=0xf5500dd0, this_binding_value=4126149459, lex_env_p=0xf5d00550, arg_list_p=0xffffcd80, arg_list_len=0)
    at jerryscript/jerry-core/vm/vm.c:4136
#24 0x5664b89b in ecma_op_function_call_simple (func_obj_p=0xf5f00700, this_arg_value=72, arguments_list_p=0xffffcd80, arguments_list_len=0)
    at jerryscript/jerry-core/ecma/operations/ecma-function-object.c:855
#25 0x5664c218 in ecma_op_function_call (func_obj_p=0xf5f00700, this_arg_value=72, arguments_list_p=0xffffcd80, arguments_list_len=0)
    at jerryscript/jerry-core/ecma/operations/ecma-function-object.c:1054
#26 0x56612842 in opfunc_call.lto_priv.446 (frame_ctx_p=0xffffcd40) at jerryscript/jerry-core/vm/vm.c:765
#27 0x565d23a3 in vm_execute (frame_ctx_p=0xffffcd40) at jerryscript/jerry-core/vm/vm.c:4034
#28 0x565d291b in vm_run (bytecode_header_p=0xf5102a30, this_binding_value=4126149459, lex_env_p=0xf5d007b0, arg_list_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:4136
#29 0x56610e9b in vm_run_global (bytecode_p=0xf5102a30) at jerryscript/jerry-core/vm/vm.c:320
#30 0x5669246e in jerry_run (func_val=4126148451) at jerryscript/jerry-core/api/jerry.c:595
#31 0x5668ea70 in main (argc=2, argv=0xffffd1d4) at jerryscript/jerry-main/main-unix.c:748
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### PR DESCRIPTION
**Pull request by rerobika**
This patch fixes #3647.

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: tests/jerry/es2015/regression-test-issue-3647.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+try {
+  f();
+  let g;
+  function f() {
+      eval('function g() {};')
```
Comment (zherczeg) [inline tests/jerry/es2015/regression-test-issue-3647.js]: I don't understand this. `g` should be in the scope of `f`, and I don't see any error here.

Comment (rerobika) [reply tests/jerry/es2015/regression-test-issue-3647.js]: Good catch, fixed.

File: jerry-core/vm/vm.c [hunk 2]
```diff
@@ -1516,36 +1516,43 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+            else
+            {
+              JERRY_ASSERT (!(*property_p & ECMA_PROPERTY_FLAG_ENUMERABLE));
+            }
+#endif /* !JERRY_NDEBUG */
           }
-#endif /* ENABLED (JERRY_ES2015) && !JERRY_NDEBUG */
 
-          result = vm_set_var (lex_env_p, name_p, is_strict, left_value);
+          result = ecma_op_set_mutable_binding (lex_env_p, name_p, left_value, is_strict);
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: `is_strict` should be false

File: jerry-core/vm/vm.c [hunk 3]
```diff
@@ -1516,36 +1516,43 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
 
-#if ENABLED (JERRY_ES2015) && !(defined JERRY_NDEBUG)
           if (ecma_get_lex_env_type (lex_env_p) == ECMA_LEXICAL_ENVIRONMENT_DECLARATIVE)
           {
             ecma_property_t *property_p = ecma_find_named_property (lex_env_p, name_p);
 
-            JERRY_ASSERT (property_p == NULL || !(*property_p & ECMA_PROPERTY_FLAG_ENUMERABLE));
+            if (property_p == NULL)
+            {
+              ecma_op_create_mutable_binding (lex_env_p, name_p, frame_ctx_p->is_eval_code);
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: Shouldn't we do a continue here? It seems we define the property twice.

Comment (rerobika) [reply jerry-core/vm/vm.c]: It did not define the property twice, only created / initialized it. I've updated the PR and simplified it.

**Pull request merged**